### ISSUE TITLE
Nodes hang to update their cluster node list

### ISSUE THREAD

**Issue by maximumG** [cilium/cilium#24574]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

We have recently updated cilium from `v1.10.8` to `v1.13.0` on our EKS cluster running in `v1.24.10`. Our setup leverages AWS spot instances and as such, nodes comes and goes in the cluster.

We started experimenting some inter-node connectivity issue while running cilium `v1.13.0`. The symptom is that pod shosted on some nodes are not able to communicate with pods hosted on some other **recently** joined nodes.

After a few investigation, we found out that a specific cilium agent running on a node was not able to update its node list. As such some old nodes were still in the `cilium status --verbose` output but with an `unreachable` status. Of course these nodes were no longer part of the cluster and were replaced by new nodes. On the other end, these new nodes were not referenced at all in the `cilium status --verbose`. As a result, pods hosted on this stuck node were not able to communicate with pods running on any new nodes.

In the meantime, the `cilium node list` on the affected node was timing out with the following message. Any other cilium command gave the expected output.

```bash
$ cilium node list
Cilium API client timeout exceeded
 ```
 
 As a workaround we deleted the cilium agent on the affected node and the newly created agent was able to catch up with the every node clusters and provided back full cluster connectivity.

### Cilium Version

1.13.0 c9723a8d 2023-02-15T14:18:31+01:00 go version go1.19.6 linux/amd64

### Kernel Version

5.10.167-147.601

### Kubernetes Version

v1.24.10-eks-48e63af

### Sysdump

I can upload the sysdump for the whole cluster, on demand if anybody is interrested. 

### Relevant log output

```shell
$ cilium node list
Cilium API client timeout exceeded

$ cilium status --verbose
KVStore:                Ok   Disabled
Kubernetes:             Ok   1.24+ (v1.24.10-eks-48e63af) [linux/amd64]
Kubernetes APIs:        ["cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumEndpoint", "cilium/v2::CiliumLocalRedirectPolicy", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "core/v1::Namespace", "core/v1::Node", "core/v1::Pods", "core/v1::Service", "discovery/v1::EndpointSlice", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:   Strict   [eth0 10.80.6.245]
Host firewall:          Disabled
CNI Chaining:           none
CNI Config file:        CNI configuration file management disabled
Cilium:                 Ok   1.13.0 (v1.13.0-c9723a8d)
NodeMonitor:            Listening for events on 2 CPUs with 64x4096 of shared memory
Cilium health daemon:   Ok   
IPAM:                   IPv4: 8/254 allocated from 100.64.4.0/24, 
Allocated addresses:
  100.64.4.107 (kube-system/node-local-dns-8r6np[restored])
  100.64.4.124 (router)
  100.64.4.125 (kube-monitoring/datadog-agent-lpw8v[restored])
  100.64.4.191 (teleport-kube-agent/teleport-kube-agent-0[restored])
  100.64.4.241 (nginx-ingress-internal/nginx-ingress-internal-ingress-nginx-controller-7c75db779dcmxp8[restored])
  100.64.4.249 (health)
  100.64.4.25 (nginx-ingress/nginx-ingress-ingress-nginx-controller-7f8fc5d6fd-8smbg[restored])
  100.64.4.28 (kube-system/ebs-csi-node-vplnz[restored])
IPv6 BIG TCP:           Disabled
BandwidthManager:       Disabled
Host Routing:           Legacy
Masquerading:           IPTables [IPv4: Enabled, IPv6: Disabled]
Clock Source for BPF:   ktime
Controller Status:      49/49 healthy
  Name                                          Last success   Last error     Count   Message
  bpf-map-sync-cilium_lxc                       5s ago         never          0       no error   
  cilium-health-ep                              33s ago        never          0       no error   
  dns-garbage-collector-job                     49s ago        never          0       no error   
  endpoint-2005-regeneration-recovery           never          never          0       no error   
  endpoint-2201-regeneration-recovery           never          never          0       no error   
  endpoint-2409-regeneration-recovery           never          never          0       no error   
  endpoint-3268-regeneration-recovery           never          never          0       no error   
  endpoint-3894-regeneration-recovery           never          never          0       no error   
  endpoint-406-regeneration-recovery            never          never          0       no error   
  endpoint-847-regeneration-recovery            never          never          0       no error   
  endpoint-958-regeneration-recovery            never          never          0       no error   
  endpoint-gc                                   2m51s ago      never          0       no error   
  ipcache-inject-labels                         2h37m26s ago   52h2m50s ago   0       no error   
  k8s-heartbeat                                 12s ago        never          0       no error   
  metricsmap-bpf-prom-sync                      5s ago         never          0       no error   
  resolve-identity-3268                         2m37s ago      never          0       no error   
  resolve-identity-406                          2m36s ago      never          0       no error   
  restoring-ep-identity (2005)                  52h2m37s ago   never          0       no error   
  restoring-ep-identity (2201)                  52h2m37s ago   never          0       no error   
  restoring-ep-identity (2409)                  52h2m37s ago   never          0       no error   
  restoring-ep-identity (3268)                  52h2m37s ago   never          0       no error   
  restoring-ep-identity (3894)                  52h2m37s ago   never          0       no error   
  restoring-ep-identity (847)                   52h2m37s ago   never          0       no error   
  restoring-ep-identity (958)                   52h2m37s ago   never          0       no error   
  sync-endpoints-and-host-ips                   1h17m34s ago   never          0       no error   
  sync-lb-maps-with-k8s-services                52h2m37s ago   never          0       no error   
  sync-policymap-2005                           25s ago        never          0       no error   
  sync-policymap-2201                           25s ago        never          0       no error   
  sync-policymap-2409                           25s ago        never          0       no error   
  sync-policymap-3268                           25s ago        never          0       no error   
  sync-policymap-3894                           25s ago        never          0       no error   
  sync-policymap-406                            25s ago        never          0       no error   
  sync-policymap-847                            25s ago        never          0       no error   
  sync-policymap-958                            25s ago        never          0       no error   
  sync-to-k8s-ciliumendpoint (2005)             5s ago         never          0       no error   
  sync-to-k8s-ciliumendpoint (2201)             5s ago         never          0       no error   
  sync-to-k8s-ciliumendpoint (2409)             6s ago         never          0       no error   
  sync-to-k8s-ciliumendpoint (3268)             5s ago         never          0       no error   
  sync-to-k8s-ciliumendpoint (3894)             5s ago         never          0       no error   
  sync-to-k8s-ciliumendpoint (406)              5s ago         never          0       no error   
  sync-to-k8s-ciliumendpoint (847)              5s ago         never          0       no error   
  sync-to-k8s-ciliumendpoint (958)              5s ago         never          0       no error   
  template-dir-watcher                          never          never          0       no error   
  waiting-initial-global-identities-ep (2005)   52h2m37s ago   never          0       no error   
  waiting-initial-global-identities-ep (2201)   52h2m37s ago   never          0       no error   
  waiting-initial-global-identities-ep (3268)   52h2m37s ago   never          0       no error   
  waiting-initial-global-identities-ep (3894)   52h2m37s ago   never          0       no error   
  waiting-initial-global-identities-ep (847)    52h2m37s ago   never          0       no error   
  waiting-initial-global-identities-ep (958)    52h2m37s ago   never          0       no error   
Proxy Status:            No managed proxy redirect
Global Identity Range:   min 256, max 65535
Hubble:                  Disabled
KubeProxyReplacement Details:
  Status:                 Strict
  Socket LB:              Enabled
  Socket LB Tracing:      Enabled
  Devices:                eth0 10.80.6.245
  Mode:                   SNAT
  Backend Selection:      Maglev (Table Size: 16381)
  Session Affinity:       Enabled
  Graceful Termination:   Enabled
  NAT46/64 Support:       Disabled
  XDP Acceleration:       Disabled
  Services:
  - ClusterIP:      Enabled
  - NodePort:       Enabled (Range: 30000-32767) 
  - LoadBalancer:   Enabled 
  - externalIPs:    Enabled 
  - HostPort:       Enabled
BPF Maps:   dynamic sizing: on (ratio: 0.002500)
  Name                          Size
  Non-TCP connection tracking   65536
  TCP connection tracking       131072
  Endpoint policy               65535
  Events                        2
  IP cache                      512000
  IP masquerading agent         16384
  IPv4 fragmentation            8192
  IPv4 service                  65536
  IPv6 service                  65536
  IPv4 service backend          65536
  IPv6 service backend          65536
  IPv4 service reverse NAT      65536
  IPv6 service reverse NAT      65536
  Metrics                       1024
  NAT                           131072
  Neighbor table                131072
  Global policy                 16384
  Per endpoint policy           65536
  Session affinity              65536
  Signal                        2
  Sockmap                       65535
  Sock reverse NAT              65536
  Tunnel                        65536
Encryption:                                               Wireguard        [cilium_wg0 (Pubkey: B06RhcrQEXD49uQvMoGEnRmzGszTmfaclaxayOUU9lo=, Port: 51871, Peers: 11)]
Cluster health:                                           9/12 reachable   (2023-03-24T17:18:13Z)
  Name                                                    IP               Node          Endpoints
  ip-10-80-6-245.eu-west-1.compute.internal (localhost)   10.80.6.245      reachable     reachable
  ip-10-80-4-101.eu-west-1.compute.internal               10.80.4.101      reachable     reachable
  ip-10-80-4-143.eu-west-1.compute.internal               10.80.4.143      unreachable   unreachable # node gone for a while
  ip-10-80-4-229.eu-west-1.compute.internal               10.80.4.229      reachable     reachable
  ip-10-80-4-67.eu-west-1.compute.internal                10.80.4.67       unreachable   unreachable # node gone for a while
  ip-10-80-5-13.eu-west-1.compute.internal                10.80.5.13       reachable     reachable
  ip-10-80-5-169.eu-west-1.compute.internal               10.80.5.169      reachable     reachable
  ip-10-80-5-220.eu-west-1.compute.internal               10.80.5.220      reachable     reachable
  ip-10-80-5-223.eu-west-1.compute.internal               10.80.5.223      reachable     reachable
  ip-10-80-6-118.eu-west-1.compute.internal               10.80.6.118      reachable     reachable
  ip-10-80-6-127.eu-west-1.compute.internal               10.80.6.127      reachable     reachable
  ip-10-80-6-96.eu-west-1.compute.internal                10.80.6.96       unreachable   unreachable # node gone for a while
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by julianwiedmann** [comment]
Hi, thanks for reporting this issue!

It sounds you are able to reproduce - have you checked already if `v1.13.1` is also affected? Can you attach a sysdump where a node is in such state?

**Comment by maximumG** [comment]
Hi @julianwiedmann ! Thanks for your feedback.

> It sounds you are able to reproduce 

Unfortunately I'm not able to reproduce the issue on demand. We faced it randomly twice after upgrading cilium to `v1.13.0`. 

> have you checked already if v1.13.1 is also affected

I already saw some bugfixes in v1.13.1 that might be related to what we experience. However I can't say for sure as the `cilium node list` hanging command is just a symptom and not the root cause of the issue. Anyway I will try to upgrade to v1.13.1 to see if the issue arise.

> Can you attach a sysdump where a node is in such state?


Here is an extract of the cluster sysdump focused on the affected node.: [cilium-bugtool-cilium-22zmr-20230324-181956.zip](https://github.com/cilium/cilium/files/11090629/cilium-bugtool-cilium-22zmr-20230324-181956.zip)

**Comment by maximumG** [comment]
We also started to see the same issue while upgrading from `v.12.5` to `v1.12.9` on Kops clusters.

Same symptom as being:

* the nodes has an outdated view of the cluster. Meaning old deleted nodes marked as unreachable while new nodes are not appearing in the `cilium status` command
* issuing `cilium node list` on the impacted agent times out

Based on our finding we think it might be related to some locking mechanism inside the code because of the following:

* issuing a `kubectl get ciliumnodes` inside the agent works as expected, meaning we get an up-to-date list of nodes, matching the `kubectl get nodes` list
* the cilium-agent has a working TCP connection towards one of the kube-apiserver (ESTABLISHED state)
* when we issue `cilium node list` command and capture the traffic from this socket, no traffic is ever sent inside the socket

**Comment by Izzette** [comment]
We encountered a similar issue in v1.12.9 and upon debugging we seem to have found a deadlock.  I'm not sure if it is related to "Unable to fetch kubernetes labels", however it does seem to be the cause of blocking `cillium node list`.

We see these suspect traces in the cilium agent:
<details>
  <summary>Node Manager (goroutine 210 [select, 226 minutes])</summary>
  
```text
goroutine 210 [select, 226 minutes]:
golang.org/x/sync/semaphore.(*Weighted).Acquire(0xc000434cd0, {0x34df1d0, 0xc00007c030}, 0x1)
	/go/src/github.com/cilium/cilium/vendor/golang.org/x/sync/semaphore/semaphore.go:60 +0x345
github.com/cilium/cilium/pkg/lock.(*SemaphoredMutex).RLock(...)
	/go/src/github.com/cilium/cilium/pkg/lock/semaphored_mutex.go:46
github.com/cilium/cilium/pkg/ipcache.(*IPCache).RLock(...)
	/go/src/github.com/cilium/cilium/pkg/ipcache/ipcache.go:140
github.com/cilium/cilium/pkg/wireguard/agent.(*Agent).UpdatePeer(0xc00013f830, {0xc004217080, 0x13}, {0xc002d81b00, 0x2c}, {0xc002c72c60, 0x10, 0x10}, {0x0, 0x0, ...})
	/go/src/github.com/cilium/cilium/pkg/wireguard/agent/agent.go:319 +0xa7
github.com/cilium/cilium/pkg/datapath/linux.(*linuxNodeHandler).nodeUpdate(0xc0003326e0, 0x0, 0xc004011b00, 0x0)
	/go/src/github.com/cilium/cilium/pkg/datapath/linux/node.go:1103 +0xe1d
github.com/cilium/cilium/pkg/datapath/linux.(*linuxNodeHandler).NodeValidateImplementation(_, {{0xc004217080, 0x13}, {0xc002c72ac0, 0x7}, {0xc0024c2c30, 0x2, 0x2}, 0xc002758438, {0x0, ...}, ...})
	/go/src/github.com/cilium/cilium/pkg/datapath/linux/node.go:1619 +0xf8
github.com/cilium/cilium/pkg/node/manager.(*Manager).backgroundSync.func1({0x34f3048?, 0xc0003326e0?})
	/go/src/github.com/cilium/cilium/pkg/node/manager/manager.go:343 +0xb8
github.com/cilium/cilium/pkg/node/manager.(*Manager).Iter(0x29f3c60?, 0xc00268be20)
	/go/src/github.com/cilium/cilium/pkg/node/manager/manager.go:186 +0xe6
github.com/cilium/cilium/pkg/node/manager.(*Manager).backgroundSync(0xc00042c680)
	/go/src/github.com/cilium/cilium/pkg/node/manager/manager.go:342 +0x44a
created by github.com/cilium/cilium/pkg/node/manager.NewManager
	/go/src/github.com/cilium/cilium/pkg/node/manager/manager.go:230 +0x6ed
```
</details><details>
  <summary>IP Identity Watcher (goroutine 847 [semacquire, 226 minutes])</summary>
  
```text
goroutine 847 [semacquire, 226 minutes]:
sync.runtime_SemacquireMutex(0xc000f14200?, 0x10?, 0xdb4c80?)
	/usr/local/go/src/runtime/sema.go:71 +0x25
sync.(*Mutex).lockSlow(0xc0003326e0)
	/usr/local/go/src/sync/mutex.go:162 +0x165
sync.(*Mutex).Lock(...)
	/usr/local/go/src/sync/mutex.go:81
github.com/cilium/cilium/pkg/datapath/linux.(*linuxNodeHandler).AllocateNodeID(0xc0003326e0, {0xc004f819f0, 0x10, 0x10})
	/go/src/github.com/cilium/cilium/pkg/datapath/linux/node_ids.go:42 +0xf6
github.com/cilium/cilium/pkg/ipcache.(*IPCache).upsertLocked(0xc0004670e0, {0xc004f81a20, 0xd}, {0xc004f819f0, 0x10, 0x10}, 0x0, 0xc001ef22a0, {0x5, {0x2f7662a, ...}, ...}, ...)
	/go/src/github.com/cilium/cilium/pkg/ipcache/ipcache.go:429 +0x1742
github.com/cilium/cilium/pkg/ipcache.(*IPCache).Upsert(0xc0004670e0, {0xc004f81a20, 0xd}, {0xc004f819f0, 0x10, 0x10}, 0x45?, 0x28?, {0x5, {0x2f7662a, ...}, ...})
	/go/src/github.com/cilium/cilium/pkg/ipcache/ipcache.go:260 +0x14d
github.com/cilium/cilium/pkg/ipcache.(*IPIdentityWatcher).Watch(0xc002024000, {0x34df1d0, 0xc00007c038})
	/go/src/github.com/cilium/cilium/pkg/ipcache/kvstore.go:318 +0x12c5
github.com/cilium/cilium/pkg/ipcache.(*IPCache).InitIPIdentityWatcher.func1.1()
	/go/src/github.com/cilium/cilium/pkg/ipcache/kvstore.go:399 +0x165
created by github.com/cilium/cilium/pkg/ipcache.(*IPCache).InitIPIdentityWatcher.func1
	/go/src/github.com/cilium/cilium/pkg/ipcache/kvstore.go:395 +0x5a
```
</details>
<details>
  <summary>Serve HTTP get_cluster_nodes (goroutine 145804 [semacquire, 135 minutes])</summary>
  
```text
goroutine 145804 [semacquire, 135 minutes]:
sync.runtime_SemacquireMutex(0xffffffffffffffff?, 0x1?, 0xc001d544b8?)
	/usr/local/go/src/runtime/sema.go:71 +0x25
sync.(*Mutex).lockSlow(0xc00489be60)
	/usr/local/go/src/sync/mutex.go:162 +0x165
sync.(*Mutex).Lock(...)
	/usr/local/go/src/sync/mutex.go:81
github.com/cilium/cilium/pkg/node/manager.(*Manager).GetNodes(0xc00042c680)
	/go/src/github.com/cilium/cilium/pkg/node/manager/manager.go:679 +0x15a
github.com/cilium/cilium/daemon/cmd.(*Daemon).getNodeStatus(0xc0010e8780)
	/go/src/github.com/cilium/cilium/daemon/cmd/status.go:419 +0x125
github.com/cilium/cilium/daemon/cmd.(*getNodes).Handle(0x4d9d038?, {0xc00023e500?, 0x0?})
	/go/src/github.com/cilium/cilium/daemon/cmd/status.go:559 +0x5bd
github.com/cilium/cilium/api/v1/server/restapi/daemon.(*GetClusterNodes).ServeHTTP(0xc001496030, {0x34cdc98, 0xc004fb4fa8}, 0xc00023e500)
	/go/src/github.com/cilium/cilium/api/v1/server/restapi/daemon/get_cluster_nodes.go:57 +0x119
github.com/go-openapi/runtime/middleware.NewOperationExecutor.func1({0x34cdc98, 0xc004fb4fa8}, 0xc00023e500)
	/go/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/operation.go:28 +0x59
net/http.HandlerFunc.ServeHTTP(0xb?, {0x34cdc98?, 0xc004fb4fa8?}, 0x4740f7?)
	/usr/local/go/src/net/http/server.go:2084 +0x2f
github.com/go-openapi/runtime/middleware.NewRouter.func1({0x34cdc98, 0xc004fb4fa8}, 0xc00023e100)
	/go/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/router.go:78 +0x257
net/http.HandlerFunc.ServeHTTP(0x203001?, {0x34cdc98?, 0xc004fb4fa8?}, 0x3?)
	/usr/local/go/src/net/http/server.go:2084 +0x2f
github.com/go-openapi/runtime/middleware.Redoc.func1({0x34cdc98, 0xc004fb4fa8}, 0x2?)
	/go/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/redoc.go:72 +0x242
net/http.HandlerFunc.ServeHTTP(0xc001dbb980?, {0x34cdc98?, 0xc004fb4fa8?}, 0x40?)
	/usr/local/go/src/net/http/server.go:2084 +0x2f
github.com/go-openapi/runtime/middleware.Spec.func1({0x34cdc98, 0xc004fb4fa8}, 0xc001dbb901?)
	/go/src/github.com/cilium/cilium/vendor/github.com/go-openapi/runtime/middleware/spec.go:46 +0x18c
net/http.HandlerFunc.ServeHTTP(0xc0043b4f40?, {0x34cdc98?, 0xc004fb4fa8?}, 0x0?)
	/usr/local/go/src/net/http/server.go:2084 +0x2f
github.com/cilium/cilium/pkg/metrics.(*APIEventTSHelper).ServeHTTP(0xc0012b0210, {0x34ddc58?, 0xc005eac7e0}, 0xc00023e100)
	/go/src/github.com/cilium/cilium/pkg/metrics/middleware.go:63 +0x1d4
github.com/cilium/cilium/pkg/api.(*APIPanicHandler).ServeHTTP(0x0?, {0x34ddc58?, 0xc005eac7e0?}, 0xc000880400?)
	/go/src/github.com/cilium/cilium/pkg/api/apipanic.go:42 +0x76
net/http.serverHandler.ServeHTTP({0xc005089230?}, {0x34ddc58, 0xc005eac7e0}, 0xc00023e100)
	/usr/local/go/src/net/http/server.go:2916 +0x43b
net/http.(*conn).serve(0xc003c6e780, {0x34df240, 0xc0019581b0})
	/usr/local/go/src/net/http/server.go:1966 +0x5d7
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:3071 +0x4db
```
</details>

It would appear that goroutine 210 and 847 are in a classic deadlock, blocking goroutine 145804 from completing.
As best I can, here is the flowchart of deadlock below.  It seems to occur when wireguard is enabled during high node-churn.

```mermaid
---
title: "IPCache.mutex ↔ linuxNodeHandler.mutex deadlock"
---
flowchart TD
    NewManager["pkg/node/manager.NewManager"] --->|"go func()"| Manager
    InitIPIdentityWatcher["pkg/ipcache.(*IPCache).InitIPIdentityWatcher.func1"] --->|"go func()"| IPIdentityWatcher

    subgraph data [" "]
        subgraph IPCache["pkg/ipcache.IPCache{}"]
            IPCacheMutex["🔐 mutex"]
        end
        subgraph NodeHandler["pkg/datapath/linux.linuxNodeHandler{}"]
            NodeHandlerMutex["🔐 mutex"]
        end
    end
    style data fill:#00000000,stroke-width:0px

    subgraph IPIdentityWatcher
        direction LR
        InitIPIdentityWatcherFunc11["
          pkg/ipcache.(*IPCache).InitIPIdentityWatcher.func1.1 ⏎
          pkg/ipcache.(*IPIdentityWatcher).Watch
        "] -->
        Upsert["pkg/ipcache.(*IPCache).Upsert"]
            <==>|"sync.(*Mutex).Lock *held*"| IPCacheMutex
        Upsert -->
        upsertLocked["
            pkg/ipcache.(*IPCache).upsertLocked ⏎
            pkg/datapath/linux.(*linuxNodeHandler).AllocateNodeID
        "] --->|"sync.(*Mutex).Lock *blocks*"| NodeHandlerMutex
    end


    subgraph Manager
        direction RL
        backgroundSync["
            pkg/node/manager.(*Manager).backgroundSync ⏎
            pkg/node/manager.(*Manager).Iter ⏎
            pkg/node/manager.(*Manager).backgroundSync.func1
        "] -->
        NodeValidateImpl["pkg/datapath/linux.(*linuxNodeHandler).NodeValidateImplementation"]
            <==>|"sync.(*Mutex).Lock *held*"| NodeHandlerMutex
        NodeValidateImpl -->
        UpdatePeer["pkg/wireguard/agent.(*Agent).UpdatePeer"]
            -..->|"
                pkg/ipcache.(*IPCache).RLock ⏎
                pkg/lock.(*SemaphoredMutex).RLock ⏎
                golang.org/x/sync/semaphore.(*Weighted).Acquire *blocks*
            "| IPCacheMutex
    end
```

I'm working on a fix now!  I hope to send a PR within the next couple of days.

**Comment by DerrickMartinez** [comment]
@Izzette  We are also running wireguard and running into "Unable to fetch kubernetes labels". Since our PolicyEnforcement=default, it was also breaking (dropping) traffic even though there are no rules. Since the evaluation of the rules (or unable to) would default to Deny. 

Also `cilium ip list` was resulting in a time exceeded.

Hopefully your fix will also help in our situation as well.

**Comment by bimmlerd** [comment]
@Izzette nicely diagnosed and explained, I agree that you've found a deadlock. I've had a look at the internals surrounding this, and saw that it might be possible to isolate the node ID handling (and locking) into a struct of it's own. That way we don't need to lock the whole `linuxNodeHandler` and avoid the cycle. Does that match your idea for a fix? Happy to review a PR, or put one up if you don't currently have the cycles.

**Comment by Izzette** [comment]
@bimmlerd actually I was thinking that the `wireguard.(*Agent).UpdatePeer`/`DeletePeer` do not need to be performed in-band nor with `linuxNodeHandler` locked.  I was struggling with the development environment on my Mac M1, and it kind of fell to the back burner.  I have my changes, but I wasn't able to write the appropriate tests.  The changes are really simple however, but I'm only 90% sure the impact is not harmful.  I intend to give up on trying to test directly on this box and use a remote Linux AMD64 machine for testing.  On our end, we're considering abandoning wireguard encryption.  If we go that route, I may never get around to it.  I've pushed what I have however: https://github.com/cilium/cilium/pull/25234

**Comment by bimmlerd** [comment]
@Izzette 

Thanks! I'll try to have a look at the PR, though maybe I'll defer to someone more knowledgable about the wireguard implementation.

> I was struggling with the development environment on my Mac M1

Yeah I get it, I'm on a M1 too. I see two options - dev in a VM (e.g. with [lima](https://github.com/lima-vm/lima), [Multipass](https://multipass.run/), ...) or running [Asahi](https://asahilinux.org/). I chose Asahi, but it's not [without pitfalls](https://github.com/AsahiLinux/docs/wiki/Feature-Support). I think it'd need a bit of effort to make cilium dev on MacOS possible, but it's hard not to break it when everyone and CI is linux-only :/

**Comment by Izzette** [comment]
Excelent, thanks !

**Comment by ravilr** [comment]
Please cherrypick fixes to 1.12 and 1.13 ? as this was a regression introduced in cilium-v1.12.8 ?
